### ISSUE TITLE
[package/fx] make custom tracers more pluggable

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#54573]
If the user traces a module with a custom FX tracer and produces a GraphModule, then packages it, things will break today, as the `__reduce_package__` on GraphModule just uses a vanilla tracer to reconstruct the model. D27278219 is an example where @bradleyhd fixes this by overriding `__reduce_package__` to use the "right" tracer.

The approach in D27278219 works fine, but ideally we should make this "just work". An idea: can we store the "originating tracer" object on the GraphModule and pickle that along with the GraphModule, then just use that to reconstruct things?

**Comment by bradleyhd** [comment]
makes sense to me to store the originating tracer on the GraphModule, it would make things super straightforward for our use case

**Comment by suo** [comment]
heads up @jamesr66a that this applies not just to packaging but any form of GraphModule serializaiton

**Comment by SplitInfinity** [comment]
Turns out `Tracer` instances are not pickable:
```
>>> import pickle                                                               
>>> import torch                                                                
>>> t = torch.fx.Tracer()                            
>>> from torch.fx import Tracer   
>>> t = Tracer()
>>> f = open("/tmp/pickle.pt", "wb")     
>>> pickler = pickle.Pickler(f)          
>>> pickler.dump(t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot pickle 'module' object 
```

This is likely due to the `_autowrap_search` attribute of `Tracer`:
https://github.com/pytorch/pytorch/blob/63d54874e7f60a2ff768131896cd44cb562f4993/torch/fx/symbolic_trace.py#L141

**Comment by suo** [comment]
I'm not sure what that functionality is for. @jamesr66a any thoughts? I think it would be good to make `Tracer` pickleable so that custom tracers will serialize correctly, anything we can do to avoid storing modules and the like on `Tracer`?

**Comment by jamesr66a** [comment]
It should be fine to serialize those as string module qualnames so long as they resolve to the correct thing on the loading process side. I'm not exactly sure what we have to do to define the semantics to ensure that it does something the user expects
